client/components/data/document-head/index.jsx

```diff
@@ -13,8 +13,7 @@ import { getDocumentHeadFormattedTitle } from 'calypso/state/document-head/selec
 import { getDocumentHeadTitle } from 'calypso/state/document-head/selectors/get-document-head-title';
 
 class DocumentHead extends Component {
-	// @TODO: Please update https://github.com/Automattic/wp-calypso/issues/58453 if you are refactoring away from UNSAFE_* lifecycle methods!
-	UNSAFE_componentWillMount() {
```

Comment:
Just a quick observation after first look: is the `DocumentHead` component rendered in SSR? Then the cWM method runs on SSR, while cDM doesn't. Can it possibly make a difference?

Reply:
Yes, there is a subtle difference. On SSR, the `DocumentHead` won't set the title immediately. However, since we hydrate the server-rendered pages, the title will be set on the client. The behavior change that's introduced by this change is that the title won't be set initially, but with a slow delay (on the client). I'm personally fine with that, since relying on a lifecycle method for SSR is not a good idea anyway.

Reply:
When, for example, the `/log-in` page is server-rendered, there's this `<DocumentHead />` instance being processed:

https://github.com/Automattic/wp-calypso/blob/trunk/client/login/wp-login/index.jsx#L293-L297

And its effects are really visible in the server-rendered HTML: there's a correct `title` and the `meta[description]` and `link[rel=canonical]` tags:

<img width="443" alt="Screenshot 2022-03-29 at 9 44 48" src="https://user-images.githubusercontent.com/664258/160561357-0cfbef2c-c3a6-44c4-bb69-5017314af638.png">

But with this patch applied, the effects are not there: there's a default `WordPress.com` title and the `meta` and `link` tags are missing. The client-side hydration fixes the title but interestingly, for some reason, it doesn't add the missing `meta` and link`.

After this patch `DocumentHead` stops working with SSR and I don't think that's acceptable. I'm afraid there's no simple fix and we'll need to rethink how `DocumentHead` works.

Reply:
> Are you sure `DocumentHead` actually works with SSR in production right now?

I finally investigated this more closely and the answer is: yes and no 🙂 It only works if the server-side render is not cached, i.e., only on the very first request. Look at this code:

https://github.com/Automattic/wp-calypso/blob/trunk/client/server/render/index.js#L222-L234

The order of events is:
1. If SSR-eligible, we call `render( context.layout, cacheKey )` to render the React tree, and that render is internally cached.
2. Then we call `attachHead` which reads the `state.documentHead` data from Redux and writes it to `context`, to be later written to the response.

But the correct `documentHead` data is present in the Redux store only if the React tree was actually rendered, and if all the `componentWillMount` methods were called. These methods write the `documentHead` data to Redux. If the render result is already cached, these writes don't happen.

After this PR, the writes will never happen because they are in methods that SSR doesn't call. But the practical difference is zero because before this PR the writes happened only on the very first render. The vast majority of renders doesn't have the desired `documentHead`.

To summarize, `DocumentHead` is supposed to work in SSR, but it doesn't, and this PR moves us even farther away from making it work.